src/window_identifier/wayland.rs

```diff
@@ -1,40 +1,79 @@
 use std::fmt;
 
+use wayland_backend::sys::client::Backend;
 use wayland_client::{
-    protocol::{__interfaces::WL_SURFACE_INTERFACE, wl_surface::WlSurface},
+    protocol::{__interfaces::WL_SURFACE_INTERFACE, wl_registry, wl_surface::WlSurface},
     Proxy, QueueHandle,
 };
-use wayland_protocols::xdg::foreign::zv2::client::{
-    zxdg_exported_v2::{Event, ZxdgExportedV2},
-    zxdg_exporter_v2::ZxdgExporterV2,
+use wayland_protocols::xdg::foreign::{
+    zv1::client::{
+        zxdg_exported_v1::{self, ZxdgExportedV1},
+        zxdg_exporter_v1::ZxdgExporterV1,
+    },
+    zv2::client::{
+        zxdg_exported_v2::{self, ZxdgExportedV2},
+        zxdg_exporter_v2::ZxdgExporterV2,
+    },
 };
 
 use super::WindowIdentifierType;
 
+// Supported versions.
+const ZXDG_EXPORTER_V1: u32 = 1;
+const ZXDG_EXPORTER_V2: u32 = 1;
+
+#[derive(Debug)]
 pub struct WaylandWindowIdentifier {
-    exported: ZxdgExportedV2,
+    exported_v1: Option<ZxdgExportedV1>,
+    exported_v2: Option<ZxdgExportedV2>,
     type_: WindowIdentifierType,
 }
 
 impl WaylandWindowIdentifier {
-    pub fn new(surface: &WlSurface) -> Option<Self> {
-        match wayland_handle_export_from_surface(surface) {
-            Ok((exported, handle)) => Some(Self {
-                exported,
-                type_: WindowIdentifierType::Wayland(handle),
-            }),
-            _ => None,
-        }
+    pub async fn new(surface: &WlSurface) -> Option<Self> {
+        let backend = surface.backend().upgrade()?;
+        let conn = wayland_client::Connection::from_backend(backend);
+
+        Self::new_inner(conn, surface).await
     }
 
-    pub fn from_raw(surface_ptr: *mut std::ffi::c_void) -> Option<Self> {
-        match wayland_handle_export(surface_ptr) {
-            Ok((exported, handle)) => Some(Self {
-                exported,
-                type_: WindowIdentifierType::Wayland(handle),
-            }),
-            _ => None,
+    pub async unsafe fn from_raw(
+        surface_ptr: *mut std::ffi::c_void,
+        display_ptr: *mut std::ffi::c_void,
+    ) -> Option<Self> {
+        if surface_ptr.is_null() || display_ptr.is_null() {
+            return None;
         }
+
+        let backend = Backend::from_foreign_display(display_ptr as *mut _);
+        let conn = wayland_client::Connection::from_backend(backend);
+        let obj_id = wayland_backend::sys::client::ObjectId::from_ptr(
+            &WL_SURFACE_INTERFACE,
+            surface_ptr as *mut _,
+        )
+        .ok()?;
+
+        let surface = WlSurface::from_id(&conn, obj_id).ok()?;
+
+        Self::new_inner(conn, &surface).await
+    }
+
+    async fn new_inner(conn: wayland_client::Connection, surface: &WlSurface) -> Option<Self> {
+        let (sender, receiver) =
+            futures::channel::oneshot::channel::<Option<WaylandWindowIdentifier>>();
+
+        // Cheap clone, protocol objects are essentially smart pointers
```

Comment:
```suggestion
        // Cheap clone, protocol objects are essentially small pointers
```
iirc

Reply:
Smart was the word used at https://matrix.to/#/#smithay:matrix.org when I asked about it.

Reply:
I think that smart pointer is the correct term here, wayland-rs objects are made of `ObjectId` and ref counted user data, so they essentially act like smart pointer.